spyder_kernels/customize/spydercustomize.py

```diff
@@ -898,52 +952,43 @@ def runfile(filename=None, args=None, wdir=None, namespace=None,
         __umr__.run()
     if args is not None and not isinstance(args, basestring):
         raise TypeError("expected a character buffer object")
-    if namespace is None:
-        if current_namespace:
-            namespace = _get_globals()
-        else:
-            main_mod = ipython_shell.new_main_mod(filename, '__main__')
-            namespace = main_mod.__dict__
-            # Needed to allow pickle to reference main
-            sys.modules['__main__'] = main_mod
-    namespace['__file__'] = filename
-    sys.argv = [filename]
-    if args is not None:
-        for arg in shlex.split(args):
-            sys.argv.append(arg)
-    if wdir is not None:
-        try:
-            wdir = wdir.decode('utf-8')
-        except (UnicodeError, TypeError, AttributeError):
-            # UnicodeError, TypeError --> eventually raised in Python 2
-            # AttributeError --> systematically raised in Python 3
-            pass
-        os.chdir(wdir)
-    if post_mortem:
-        set_post_mortem()
-    if __umr__.has_cython:
-        # Cython files
-        with io.open(filename, encoding='utf-8') as f:
-            ipython_shell.run_cell_magic('cython', '', f.read())
-    else:
-        execfile(filename, namespace)
-
-    ipython_shell.user_ns.update(namespace)
-
-    clear_post_mortem()
-    sys.argv = ['']
 
-    try:
-        del sys.modules['__main__']
-    except KeyError:
-        pass
-
-    # Avoid error when running `%reset -f` programmatically
-    # See issue spyder-ide/spyder-kernels#91
-    try:
-        namespace.pop('__file__')
-    except KeyError:
-        pass
+    with FileNamespace(filename, namespace, current_namespace) as namespace:
+        sys.argv = [filename]
+        if args is not None:
+            for arg in shlex.split(args):
+                sys.argv.append(arg)
+        if wdir is not None:
+            try:
+                wdir = wdir.decode('utf-8')
+            except (UnicodeError, TypeError, AttributeError):
+                # UnicodeError, TypeError --> eventually raised in Python 2
+                # AttributeError --> systematically raised in Python 3
+                pass
+            os.chdir(wdir)
+        if post_mortem:
+            set_post_mortem()
+
+        if __umr__.has_cython:
+            # Cython files
+            with io.open(filename, encoding='utf-8') as f:
+                ipython_shell.run_cell_magic('cython', '', f.read())
+        else:
+            try:
+                execfile(filename, namespace)
+            except SystemExit as status:
+                # ignore exit(0)
+                if status.code:
+                    ipython_shell.showtraceback(exception_only=True)
+            except bdb.BdbQuit:
+                if not is_pdb:
+                    ipython_shell.showtraceback(tb_offset=2)
+            except BaseException:
+                # tb offset is 1 because we wrap execfile
```

Comment:
Could you better describe here what part of the traceback we're discarding here?

Reply:
Sure, we discard the call to `execfile` and the call to `exec`

Reply:
Ok, then please that explanation instead of the comment in this line.